src/target/at32f43x.c

```diff
@@ -364,3 +388,150 @@ static bool at32f43_mass_erase(target_s *target)
+		AT32F43x_FLASH_CTRL_USDERS | AT32F43x_FLASH_CTRL_USDULKS | AT32F43x_FLASH_CTRL_ERSTR);
+
+	return at32f43_flash_busy_wait(target, 0, NULL);
+}
+
+static bool at32f43_option_write_erased(target_s *const target, const size_t offset, const uint16_t value)
+{
+	if (value == 0xffffU)
+		return true;
+
+	at32f43_flash_clear_eop(target, 0);
+
+	/* Enable writing User System Data */
+	target_mem32_write32(target, AT32F43x_FLASH_CTRL, AT32F43x_FLASH_CTRL_USDPRGM | AT32F43x_FLASH_CTRL_USDULKS);
+
+	const uint32_t addr = AT32F43x_USD_BASE + (offset * 2U);
+	target_mem32_write16(target, addr, value);
+
+	const bool result = at32f43_flash_busy_wait(target, 0, NULL);
+	if (result || offset != 0U)
+		return result;
+
+	/* For error on offset 0, that is RDP byte, signal back the failure to erase RDP (?) */
+	const uint8_t status =
+		target_mem32_read32(target, AT32F43x_FLASH_STS) & (AT32F43x_FLASH_STS_PRGMERR | AT32F43x_FLASH_STS_EPPERR);
+	return status == AT32F43x_FLASH_STS_PRGMERR;
+}
+
+#define STM32F1_OB_COUNT     8U
+#define AT32F43x_2K_OB_COUNT 256U
+#define AT32F43x_4K_OB_COUNT 2048U
+
+static bool at32f43_option_write(target_s *const target, const uint32_t addr, const uint16_t value)
+{
+	/* Arterytek F435/F437 has either 512 bytes or 4 KiB worth of USD */
+	const target_flash_s *target_flash = target->flash;
+	const uint16_t ob_count = (target_flash->blocksize == 4096U) ? AT32F43x_4K_OB_COUNT : AT32F43x_2K_OB_COUNT;
+
+	const uint32_t index = (addr - AT32F43x_USD_BASE) >> 1U;
+	/* If this underflows, then address is out of USD range */
+	if (index > ob_count - 1U)
+		return false;
+
+	bool erase_needed = false;
+	uint16_t opt_val_single = target_mem32_read16(target, AT32F43x_USD_BASE + addr);
+	/* No change pending */
+	if (opt_val_single == value)
+		return true;
+	/* Check whether erase is needed */
+	if (opt_val_single != 0xffffU)
+		erase_needed = true;
+	/* Flip single pair-of-bytes from 0xffff to desired value and exit */
+	if (!erase_needed) {
+		return at32f43_option_write_erased(target, index, value);
+	}
+
+	uint16_t *opt_val = calloc(ob_count, sizeof(uint16_t));
+	if (!opt_val) {
+		DEBUG_ERROR("calloc: failed in %s\n", __func__);
+		return false;
+	}
+	/* Save current values */
+	for (size_t i = 0U; i < ob_count * 2U; i += 4U) {
+		const size_t offset = i >> 1U;
+		uint32_t val = target_mem32_read32(target, AT32F43x_USD_BASE + i);
+		opt_val[offset] = val & 0xffffU;
+		opt_val[offset + 1U] = val >> 16U;
+	}
+	/* Update requested entry locally */
+	opt_val[index] = value;
+	/* Wipe everything and write back. Writing matching values without an erase raises a PRGMERR. */
+	bool result = true;
+	do {
+		if (!at32f43_option_erase(target)) {
+			result = false;
+			break;
+		}
+		/* Write changed values using 16-bit accesses. */
+		for (size_t i = 0U; i < ob_count; ++i) {
+			if (!at32f43_option_write_erased(target, i, opt_val[i])) {
+				result = false;
+				break;
+			}
+		}
+	} while (false);
```

Comment:
Is there a particular reason this is wrapped in a `do`-`while (false)`? We don't mind if there is a good reason, it should just be noted because this pattern is otherwise extremely surprising and eyebrow-raising.

Reply:
It used to early-return, now I can't leak the calloc pointer. Using goto cleanup for free sounded worse.

Reply:
This smells like it should be refactored to become a full fat function then, but as this works and we're more interested at the moment in getting this support merged, we'll let that be as it is.

Reply:
It could be a real function. This idiom is effectively an anonymous function here, a poor man's lambda from C++ (or small try-catch exception replacement). For something like a string command parser it's more justified. Here I'd need to pass `*target`, `*opt_val`, `ob_count`, and return true. It could work. But OTOH I haven't tested whether calloc for 4096 flies on stlink/swlink with smaller heap (mitigate with PR1529).

Reply:
Inserted commit with suggested change. It doesn't matter because everything here is marked `static inline` and at -Os inlined properly. Please tell if I should squash it into the previous one for a cleaner history.

Reply:
Looks good - much cleaner and easier to understand 👍🏼.